cmd/find-circles/main.go

```diff
@@ -0,0 +1,78 @@
+// What it does:
+//
+// This example shows how to find circles in an image using Hough transform.
+//
+// How to run:
+//
+// 		go run ./cmd/find-circles/main.go ./images/circles.jpg
+//
+// +build example
+
+package main
+
+import (
+	"fmt"
+	"image"
+	"image/color"
+	"os"
+
+	"gocv.io/x/gocv"
+)
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Println("How to run:\n\tfind-circles [imgfile]")
+		return
+	}
+
+	filename := os.Args[1]
+
+	window := gocv.NewWindow("detected circles")
+	defer window.Close()
+
+	img := gocv.IMRead(filename, gocv.IMReadGrayScale)
+	defer img.Close()
+
+	gocv.MedianBlur(img, img, 5)
+
+	cimg := gocv.NewMat()
+	defer cimg.Close()
+
+	gocv.CvtColor(img, cimg, gocv.ColorGrayToBGR)
+
+	circles := gocv.NewMat()
+	defer circles.Close()
+
+	gocv.HoughCircles(
+		img,
+		circles,
+		gocv.HoughGradient,
+		1, // dp
+		float64(img.Rows()/8), // minDist
+		75, // param1
+		20, // param2
+		10, // minRadius
+		0,  // maxRadius
+	)
+
+	blue := color.RGBA{0, 0, 255, 0}
+	red := color.RGBA{255, 0, 0, 0}
+
+	ch := circles.Channels()
+	for i := 0; i < circles.Cols(); i++ {
+		x := int(circles.GetFloatAt(0, i*ch))
+		y := int(circles.GetFloatAt(0, i*ch+1))
+		r := int(circles.GetFloatAt(0, i*ch+2))
```

Comment:
Maybe we could create a data structure for this, like `Vec3f` in opencv, and we could have a new `GetVecFAt()` func to return a vector of floats.

```
// VecF is a generic vector of floats.
type VecF []float32
```

WDYT?

Reply:
Here is the diff:

```diff
diff --git a/cmd/find-circles/main.go b/cmd/find-circles/main.go
index a39c1af..7c7d2bc 100644
--- a/cmd/find-circles/main.go
+++ b/cmd/find-circles/main.go
@@ -58,11 +58,11 @@ func main() {
 	blue := color.RGBA{0, 0, 255, 0}
 	red := color.RGBA{255, 0, 0, 0}
 
-	ch := circles.Channels()
 	for i := 0; i < circles.Cols(); i++ {
-		x := int(circles.GetFloatAt(0, i*ch))
-		y := int(circles.GetFloatAt(0, i*ch+1))
-		r := int(circles.GetFloatAt(0, i*ch+2))
+		v := circles.GetVecfAt(0, i)
+		x := int(v[0])
+		y := int(v[1])
+		r := int(v[2])
 
 		gocv.Circle(cimg, image.Pt(x, y), r, blue, 2)
 		gocv.Circle(cimg, image.Pt(x, y), 2, red, 3)
diff --git a/core.go b/core.go
index 0806e45..073b769 100644
--- a/core.go
+++ b/core.go
@@ -1070,6 +1070,22 @@ type KeyPoint struct {
 	Octave, ClassID       int
 }
 
+// Vecf is a generic vector of floats.
+type Vecf []float32
+
+// GetVecfAt returns a vector of floats. Its size corresponds to the number of
+// channels of the Mat.
+func (m *Mat) GetVecfAt(row int, col int) Vecf {
+	ch := m.Channels()
+	v := make(Vecf, ch)
+
+	for c := 0; c < ch; c++ {
+		v[c] = m.GetFloatAt(row, col*ch+c)
+	}
+
+	return v
+}
+
 func toByteArray(b []byte) C.struct_ByteArray {
 	return C.struct_ByteArray{
 		data:   (*C.char)(unsafe.Pointer(&b[0])),
```

Reply:
I had initially looked at an implementation like you suggest @willdurand for `Vec3f` however did not keep going with it. There are quite a few different vector types used within OpenCV for different filters and algorithms, so at the time I elected to use the `GetXXXAt()` and `SetXXXAt()` functions to avoid what at the time seemed like would require a lot of code duplication. We would need 2, 3 and 4-d versions for all of the OpenCV unsigned/signed ints and various sized floats.

That said, it is still a finite set, and there are many advantages to reducing the code needed for the consumer of the package. Anyone else want to comment here? I'd like to hear more viewpoints here, if possible.

Reply:
I could open an issue for that then, or a separate PR to avoid blocking this one.

As you can see in the diff above, the `GetVecfAt()` returns a vector of floats whose size depends on the number of channels, therefore it should be flexible enough. Of course, we would need `GetVecdAt()` for `double/flaot64`, `GetVeciAt()` for `int`, etc.

Reply:
One disadvantage that occurs to me, is that it requires copying all of the data from the entire `Mat` into the `Vecf` vs. the current implementation that just accesses the original data.

Reply:
right, although it provides a simple function to access the data, which can be used for less critical apps. It took me a while to figure out how to retrieve the data for the example code (because all SO/Internet answers talk about a vector or `Vec3f`).

Reply:
I agree that a separate issue/PR would be best for this vector conversation.